src/function/scalar/date/date_diff.cpp

```diff
@@ -0,0 +1,333 @@
+#include "duckdb/function/scalar/date_functions.hpp"
+#include "duckdb/common/enums/date_part_specifier.hpp"
+#include "duckdb/common/exception.hpp"
+#include "duckdb/common/types/date.hpp"
+#include "duckdb/common/types/interval.hpp"
+#include "duckdb/common/types/time.hpp"
+#include "duckdb/common/types/timestamp.hpp"
+#include "duckdb/common/vector_operations/ternary_executor.hpp"
+#include "duckdb/common/vector_operations/vector_operations.hpp"
+#include "duckdb/common/string_util.hpp"
+#include "duckdb/storage/statistics/numeric_statistics.hpp"
+
+namespace duckdb {
+
+// This function is an implementation of the "period-crossing" date difference function from T-SQL
+// https://docs.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql?view=sql-server-ver15
+struct DateDiff {
+	static constexpr const int32_t SECS_PER_WEEK = Interval::SECS_PER_DAY * Interval::DAYS_PER_WEEK;
+	static constexpr const int32_t SECS_PER_HOUR = Interval::SECS_PER_MINUTE * Interval::MINS_PER_HOUR;
+
+	static constexpr const int64_t NANOS_PER_MICRO = 1000;
+	static constexpr const int64_t NANOS_PER_MSEC = NANOS_PER_MICRO * Interval::MICROS_PER_MSEC;
+	static constexpr const int64_t MICROS_PER_SEC = Interval::MICROS_PER_MSEC * Interval::MSECS_PER_SEC;
+	static constexpr const int64_t MICROS_PER_MIN = MICROS_PER_SEC * Interval::SECS_PER_MINUTE;
+	static constexpr const int64_t MICROS_PER_HOUR = MICROS_PER_MIN * Interval::MINS_PER_HOUR;
+
+	struct YearOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA startdate, TB enddate) {
+			return Date::ExtractYear(enddate) - Date::ExtractYear(startdate);
+		}
+	};
+
+	struct MonthOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA startdate, TB enddate) {
+			return (Date::ExtractYear(enddate) * 12 + Date::ExtractMonth(enddate) - 1) -
```

Comment:
Can we use Date::Convert here? ExtractMonth calls Date::Convert internally.

Reply:
Duh, one call instead of two. You would think I've never implement this function before...